File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:


        def counting(nums,k):

            count=0
            l,r=0,0
            n=len(nums)
            d={}
            while r<n:

                d[nums[r]]= d.get(nums[r],0)+1
                r+=1
                while len(d)>k:
                    d[nums[l]]-=1
                    if d[nums[l]]==0:
                        del d[nums[l]]
                    l+=1
                count+=(r-l+1)
            
            return count
        return counting(nums,k)-counting(nums,k-1)

        

File: 1034-subarrays-with-k-different-integers/subarrays-with-k-different-integers.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraysWithKDistinct = function(nums, k) {
    function counting(k) {
        let l = 0;
        let r = 0;
        const n = nums.length;
        let dict = {};
        let uniqueCount = 0; 
        let count = 0;

        while (r < n) {

            if (!dict[nums[r]]) {
                dict[nums[r]] = 0;
                uniqueCount++;
            }
            dict[nums[r]]++;
            r++;

            while (uniqueCount > k) {
                dict[nums[l]]--;
                if (dict[nums[l]] === 0) {
                    delete dict[nums[l]];
                    uniqueCount--;
                }
                l++;
            }

            count += r - l+1;
        }

        return count;
    }

    
    return counting(k) - counting(k - 1);
};